engines/e_padlock.c

```diff
@@ -102,16 +114,19 @@ static int padlock_bind_helper(ENGINE *e)
 
     /* Generate a nice engine name with available features */
     BIO_snprintf(padlock_name, sizeof(padlock_name),
-                 "VIA PadLock (%s, %s)",
+                 "VIA PadLock (%s, %s, %s)",
                  padlock_use_rng ? "RNG" : "no-RNG",
-                 padlock_use_ace ? "ACE" : "no-ACE");
+                 padlock_use_ace ? "ACE" : "no-ACE",
+				 padlock_use_ccs ? "CCS" : "no-CCS");
 
     /* Register everything or return with an error */
     if (!ENGINE_set_id(e, padlock_id) ||
         !ENGINE_set_name(e, padlock_name) ||
         !ENGINE_set_init_function(e, padlock_init) ||
         (padlock_use_ace && !ENGINE_set_ciphers(e, padlock_ciphers)) ||
-        (padlock_use_rng && !ENGINE_set_RAND(e, &padlock_rand))) {
+        (padlock_use_rng && !ENGINE_set_RAND(e, &padlock_rand)) ||
+        (padlock_use_ccs && !ENGINE_set_ciphers(e, zx_ciphers)) ||
+        (padlock_use_ccs && !ENGINE_set_digests(e, gmi_digests))) {
```

Comment:
The zx_ciphers will include all padlock_ciphers and sm3/sm4.

Reply:
It does override completely, but do note that this is done conditionally, all depending on the value of `padlock_use_ccs`, which I think is fine in itself, BUT is a bit confusing.

I assume that the Zhaoxin platform guarantees that the padlock ciphers will always be available, i.e. that `padlock_capability()` will always return a value with the ACE bits set.  If not, it may be that the mechanism to select what ciphers are available must happen in the ciphers function itself.

Reply:
You're right. ZX-D CPU will guarantees that the padlock ciphers will always be available.

Reply:
In that case, this is acceptable.  Would you mind adding a comment above the new `ENGINE_set_ciphers` line, saying something like this?

``` C
        /* The zx ciphers will always include all padlock ciphers as well */
```

Reply:
If the ACE bits are always set, I would prefer avoiding duplication of code given that `zx_ciphers` just adds a few more cases to the switch statement and returns an expanded list.

My suggestion would be to have 2 conditional checks on `padlock_use_ccs` in `padlock_ciphers()`:

1. to return either `padlock_cipher_nids` or `zx_cipher_nids`
2. in the `default` case of `padlock_ciphers` check for CCS support and conditionally include a nested switch statement for the additional ciphers.

This would be my recommendation, but I would like to hear opinions from the other reviewers before asking @bubble932 to implement it.